### PR TITLE
Consuming the new APIs provided by go-control-plane

### PR DESCRIPTION
**Pull request by vaamarnath**
- Consumes the changes in envoyproxy/go-control-plane#39.
- Changed Gopkg.lock to pick the latest go-control-plane.
- Changed code as per the changes in the structures
  defined in go-control-plane.
- Updated the unit tests.

Fixes issue #225 

Signed-off-by: Amarnath V. A. <vaamarnath@gmail.com>

PS: Being a new to open source contribution, please bear with the mistakes in the PR.

### CODE REVIEW

**Comment by davecheney** [comment]
Hi @vaamarnath,

thanks for your PR. I'm going to need you to do a few things

1. You need to add `signed-off-by:` to each commit in this PR. The easiest way would be to do `git commit -a --amend -s`, then `git push -f`.
2. Once you've done that, please rebase the commit against master and address the conflicts.

**Comment by davecheney** [comment]
Also, please edit the PR to use the magic words

Fixes issue #225

This is needed to keep github happy

**Comment by vaamarnath** [comment]
@davecheney I believe I have incorporated the changes as mentioned by you in the previous comments.

File: Dockerfile [hunk 1]
```diff
@@ -9,6 +9,8 @@ COPY cmd cmd
 COPY internal internal
 RUN CGO_ENABLED=0 GOOS=linux go install -ldflags="-w -s" -v github.com/heptio/contour/cmd/contour
 
+RUN go test github.com/heptio/contour/...
```
Comment (davecheney) [inline Dockerfile]: please revert this

File: internal/contour/cache.go [hunk 2]
```diff
@@ -17,7 +17,8 @@ import (
 	"sort"
 	"sync"
 
-	v2 "github.com/envoyproxy/go-control-plane/api"
+	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
```
Comment (davecheney) [inline internal/contour/cache.go]: renaming the import isn't necessary any more

File: internal/contour/cache.go [hunk 3]
```diff
@@ -235,23 +236,23 @@ func (l listenersByName) Len() int           { return len(l) }
 func (l listenersByName) Swap(i, j int)      { l[i], l[j] = l[j], l[i] }
 func (l listenersByName) Less(i, j int) bool { return l[i].Name < l[j].Name }
 
-// VirtualHostCache is a thread safe, atomic, copy on write cache of v2.VirtualHost objects.
+// VirtualHostCache is a thread safe, atomic, copy on write cache of envoy_api_v2_route.VirtualHost objects.
 type virtualHostCache struct {
 	sync.Mutex
-	values []*v2.VirtualHost
+	values []envoy_api_v2_route.VirtualHost
```
Comment (davecheney) [inline internal/contour/cache.go]: You've moved from a slice of pointers to a slice of values, please make this
```
values []*envoy_api_v2_route.VirtualHost
```

File: internal/contour/cache_test.go [hunk 4]
```diff
@@ -361,38 +362,22 @@ func TestVirtualHostCacheValuesReturnsACopyOfItsInternalSlice(t *testing.T) {
 	if &v1[0] == &v2[0] {
 		// the address of the 0th element of the values slice should not be the same
 		// if it is, then we don't have a copy.
-		t.Fatalf("VirtualHostCache, consecutive calls to Values return the same backing slice: got: %p, want: %p", &v1[0], &v2[0])
-	}
-}
-
-func TestVirtualHostCacheValuesReturnsTheSameContents(t *testing.T) {
```
Comment (davecheney) [inline internal/contour/cache_test.go]: Why did you delete this test?

File: internal/contour/listener.go [hunk 5]
```diff
@@ -267,8 +270,8 @@ func tlscontext(secret *v1.Secret, alpnprotos ...string) *v2.DownstreamTlsContex
 	}
 }
 
-func httpfilter(routename string) *v2.Filter {
-	return &v2.Filter{
+func httpfilter(routename string) envoy_api_v2_listener.Filter {
```
Comment (davecheney) [inline internal/contour/listener.go]: Please return a *envoy_api_v2_listener.Filter

File: internal/contour/listener.go [hunk 6]
```diff
@@ -310,14 +313,14 @@ func httpfilter(routename string) *v2.Filter {
 	}
 }
 
-func filterchain(useproxy bool, filters ...*v2.Filter) *v2.FilterChain {
-	fc := v2.FilterChain{
+func filterchain(useproxy bool, filters ...envoy_api_v2_listener.Filter) envoy_api_v2_listener.FilterChain {
```
Comment (davecheney) [inline internal/contour/listener.go]: Can this return a `*envoy_api_v2_listener.FilterChain`

File: internal/contour/listener_test.go [hunk 7]
```diff
@@ -135,14 +136,14 @@ func TestRecomputeListener(t *testing.T) {
 
 func TestRecomputeTLSListener(t *testing.T) {
 	ingresss_http := listener(ENVOY_HTTPS_LISTENER, "0.0.0.0", 8443)
-	ingresss_http.FilterChains = []*v2.FilterChain{{
-		Filters: []*v2.Filter{
+	ingresss_http.FilterChains = []envoy_api_v2_listener.FilterChain{{
```
Comment (davecheney) [inline internal/contour/listener_test.go]: I think this unintentional change from a slice of pointers to a slice of values has caused a lot of unnecessary churn.

File: internal/contour/listener.go (hunk 6 above)
Comment (vaamarnath) [reply internal/contour/listener.go]: I see that the struct Listener type has changed it's struct field "Address" from
```go
Address *Address `protobuf:"bytes,2,opt,name=address" json:"address,omitempty"`
```
to
```go
Address envoy_api_v2_core.Address `protobuf:"bytes,2,opt,name=address" json:"address"`
```.

Is there a better way to convert the slice of pointers that `contour` is using to the slice of values that `envoy`'s APIs are expecting?

**Comment by vaamarnath** [comment]
Envoy's `go-data-plane` has changed the API locations as well as the struct types internally which resulted in converting the slice of pointers into slice of values as expected by `go-data-plane`. Tbh, I am new to `golang` and I was not quite sure myself if I had to go through pain of converting the slices of pointers to slices of values. I could not find a better way to incorporate the changes to consume the changes in APIs.

**Comment by davecheney** [comment]
Thanks. I'll have another look this afternoon and cross check with the
updated protobufs.

**Comment by sevein** [comment]
@vaamarnath, I've found that all dependencies have been updated to the latest versions allowed by `Gopkg.toml`. You could use `dep ensure -update github.com/envoyproxy/go-control-plane` instead to update just that package alone.

Related: https://golang.github.io/dep/docs/daily-dep.html#updating-dependencies.

File: internal/contour/cache_test.go (hunk 4 above)
Comment (vaamarnath) [reply internal/contour/cache_test.go]: I had make a slight change in the implementation of `virtualHostCache` and I felt that this unit test is not relevant with the changes I made.

File: internal/grpc/grpc.go [hunk 8]
```diff
@@ -241,7 +244,7 @@ type RDS struct {
 // Resources returns the contents of RDS"s cache as a []*types.Any.
 // TODO(dfc) cache the results of Resources in the VirtualHostCache so
 // we can avoid the error handling.
-func (r *RDS) Resources() ([]*types.Any, error) {
```
Comment (davecheney) [inline internal/grpc/grpc.go]: you should not have to change the signature of this function.

Comment (vaamarnath) [reply internal/grpc/grpc.go]: @davecheney I have sticked to not changing the signature as per your suggestion in [commit](https://github.com/heptio/contour/pull/226/commits/f0456317e957b5332cdb792782e4b4ea678cfe21). To achieve this, I had to have a `for` loop to convert the `[]*type.Any` to `[]type.Any` type.

If this is the acceptable approach, I can do the same for the entire PR. I will hold on to get a go ahead from you so that I am sure that I am not making some naive mistake of going down a wrong path.

**Comment by davecheney** [comment]
@vaamarnath thank you for your work. I'll use this PR as a starting point and take it from here.
